### DISCUSSION TITLE
Insert recurring transaction before due date

### DISCUSSION THREAD

**Discussion by zynexiz** [firefly-iii/firefly-iii#8759] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/explanation/more-information/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

As far as I know, the recurring transactions is inserted in the database on the due date. Sometimes you want to see the transaction before the due date (the due date can vary by some days in the bank). I know this feature was in Skrooge that I used before migrating to Frefly, and I always had the transaction inserted 10 days before.

That way I know that a payment is in the process, and for me it's a good way to check so the balance on my account is enough before the actual transaction is done in my bank.

### Solution

Could be an option in the settings (global) so it process the recurring transactions X days before the due date, like from 0 days (due date) or X days before that.

### What are alternatives?

Nothing. The recurring transaction is added on due date. I can manually "trigger" it the rule, but that's not the same thing.

**Comment by JC5** [comment]
You mean like, always create it X days before the due date?

**Comment by zynexiz** [comment]
> You mean like, always create it X days before the due date?

Yeh, precisely. Like if a transaction is due on the 28th, it will be written to the database on the 18th (if set to 10 days), of course with the original date :) An alternative is to set that configuration for each recurring transaction (then you could set different offsets), but I think one setting that applies globally is good enough?

**Comment by JC5** [comment]
Mmm. Kind of a niche feature, not gonna lie. I'll think about it.

**Comment by JC5** [comment]
Wait, no need. You can do this by making the cron job pretend it's X days in the future.

https://docs.firefly-iii.org/how-to/firefly-iii/advanced/cron/#extra-information

**Reply by zynexiz** [reply, marked as answer]
Ah, must have missed that. I will try that implementation. It will not affect any other parts in a negative way doing this?

[EDIT]
Added `/usr/bin/php /srv/http/firefly-iii/artisan firefly:cron --force --date=$(date '+\%C\%y-\%m-\%d' -d "+10 days")` to cron jobb. Just created a dummy recurring transaction on the 16th, and manually started the cron jobb. The transaction got added without issue today, so that solution worked like a charm :+1:

**Reply by JC5** [reply]
Nice work! You should be fine in regards to the side effects :)

**Reply by zynexiz** [reply]
Just thought about something, I missed this solution when I tried looking for it. Maybe it should be moved to the part in the documentation from _Extra information_ to _Calling a command_ instead? I didn't read further because the documentation started talking about Docker and IFTTT after that, so it's very easy to miss now.

**Reply by jogojapan** [reply]
Is this also possible when running Firefly-III in a docker container? The standard setup is that there is another container that runs a cronjob and calls an API over the network:

    curl https://demo.firefly-iii.org/api/v1/cron/[token]

How would one add the "future date" information in this case?

**Reply by jogojapan** [reply]
OK, looking at [the source](https://github.com/firefly-iii/firefly-iii/blob/main/app/Api/V1/Requests/System/CronRequest.php), it seems you can add it as a parameter:

    https://demo.firefly-iii.org/api/v1/cron/[token]?force=true&date=2024-04-16

And I tested it, it works.

But I'd have to say this approach for entering recurring transactions before their due-date has a few downsides.

Firstly, you have to run it for every day in the future that you expect a transaction to occur on. So you cannot, e.g., run it for a day 1 month in the future and hope that all recurring transactions between now and 1 month from now will be added. Instead you'd have to run it for every day between now and in 1 month, i.e. about 30 times.

Secondly, if you create a new recurring transaction and then trigger the "cron" API, it will not only add the new transaction, but also any other pre-existing recurring transaction that would trigger on the same day, even if that has been added already, so it will create duplicate entries.

I am also migrating from Skrooge, and this is a feature that I miss. I am not sure if I can work with Firefly-III without finding some kind of work-around.

**Reply by JC5** [reply]
Yes, this is [documented](https://api-docs.firefly-iii.org/#/about/getCron) in the API documentation.

> Firstly, you have to run it for every day in the future that you expect a transaction to occur on. So you cannot, e.g., run it for a day 1 month in the future and hope that all recurring transactions between now and 1 month from now will be added. Instead you'd have to run it for every day between now and in 1 month, i.e. about 30 times.

I'm not sue what you're trying to say. 

> Secondly, if you create a new recurring transaction and then trigger the "cron" API, it will not only add the new transaction, but also any other pre-existing recurring transaction that would trigger on the same day, even if that has been added already, so it will create duplicate entries.

Indeed, since you also seem to have set `force=true`. This is documented behavior.

**Reply by jogojapan** [reply]
Thanks for the API documentation link.

> I'm not sue what you're trying to say.

Sorry for not being clear. I tested a recurring transaction that was configured to trigger on the 7th of every month. So the next occurrence would be 2025-05-07. When I ran the cron job with `date=2025-05-31` (the end of the month), it did not create a transaction for 2025-05-07. But it did when I ran it with `date=2025-05-07`. So apparently, it views recurring transactions per-day, i.e. only creates transactions that trigger on the specific day that it considers as "today".

The Skrooge feature I was referring to is its "scheduled operations": When you define a scheduled operation, you can define how many days into the future you want it to pre-create transactions. For example, you can set "1 month", and then it will enter future transactions for that 1 month right away (it permits a maximum of 99 days into the future). And whenever you start Skrooge, it will check whether any new occurrences need to be auto-created, based on how much time has passed since the last time it ran that particular scheduled operation. (It tracks that for each scheduled operation separately.)

From what I see with the recurring transactions in Firefly-III—forgive me if I am missing anything—, they are only generated for the day that we define as "today", and not for every day between the last time it ran and that day.

So I figured that, theoretically, I could emulate the desired effect by running it once for every day between now and the desired future day. But if that day is 30 days in the future, it would have to make the `cron` API call 30 times, and indeed use the `force=true` parameter.

**Reply by JC5** [reply]
> From what I see with the recurring transactions in Firefly-III—forgive me if I am missing anything—, they are only generated for the day that we define as "today", and not for every day between the last time it ran and that day.

Yes, because that is what the cron job is based on, to be run every day. That's the design. Running it 30 times in a row to simulate a month (ahead, I guess?) isn't the use case.

**Reply by jogojapan** [reply]
Right.

I suppose you'd not consider something similar to the "scheduled operations" as a possible feature for Firefly-III?

It's similar to the recurring transactions you have already, but it would track, for each separately, how many days into the future it has been processed, and then the cronjob would fill the transactions between that day and the day defined as "today".

I noticed that on other threads you commented how you don't see forecasting as a feature for Firefly-III. This specific feature would be predictive, but not in the sense of estimation or extrapolation from the past. It would just enable auto-creation of specific, known and configured transactions.

Firefly-III already allows us to see "reports" for future time periods, which is why I actually thought it would reflect recurring transactions already there.

**Reply by JC5** [reply]
No, sorry.
